Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm

```diff
@@ -3938,6 +3941,15 @@ - (CGFloat)_overrideDeviceScaleFactor
     return _page->customDeviceScaleFactor().value_or(0);
 }
 
++ (void)_permissionChanged:(NSString *)permissionName forOrigin:(WKSecurityOrigin *)origin
+{
+    auto name = WebCore::Permissions::toPermissionName(permissionName);
+    if (!name)
+        return;
+
+    WebKit::WebProcessProxy::permissionChanged(*name, origin->_securityOrigin->securityOrigin());
+}
+
```

Comment:
I see a problem with this design: It's process-global. 

In our software, a process is typically composed of code written by many different people and teams. There's application code (and the application may be split into many parts), and also code from different frameworks.

This means that, within a given process, there may be many WebProcesses serving many different purposes. A person invoking a process-global API may trigger effects on WebViews that they didn't expect and weren't even aware of. As a result, process-global API almost always causes some sort of problem in the end. 

At a minimum, this design is inefficient. For example, it will send IPC to the Foundation NSAttributed string WebProcess, to cached, back-forward cached, pre-loaded, and pre-warmed web processes, to extensions WebProcesses, and so on when Safari changes permissions -- for no reason. In Chrome on iOS, a change in an Incognito WebProcess's settings will send IPC a standard WebProcess, even though they don't share settings.

At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.

I think we should reformulate this API to be per-WebView, and have the client app invoke the API on the WebViews it wishes to include.

Reply:
> At a minimum, this design is inefficient.

True, we might optimize it if UI process sends message to only web processes that has queried permission or has permission observers. 

> At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.

Not sure why this would happen. What are they trying to assert or check? I think the messages are handled by WebKit. 

> I think we should reformulate this API to be per-WebView, and have the client app invoke the API on the WebViews it wishes to include.

I think the reason we don't do this is client cannot notify SharedWorker and ServiceWorker about the change. 
For permission in these workers, we pick the most reasonable WebView to query permission. If we take WebView as parameter, when the permission changes for the WebView, we don't know if we should update these workers.

Reply:
> True, we might optimize it if UI process only sends message to web processes that has queried permission or has permission observers.

This is still inefficient in a process where multiple WebViews from distinct frameworks all make use of permissions. To the extent that it might wake up a process (or 100!) that was otherwise sleeping, the inefficiency may sometimes be significant.

> > At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.
> Not sure why this would happen. What are they trying to assert or check? I think the messages are handled by WebKit.

This behavior is observable to a client because it will sometimes call out to the client to ask for permission status. Right?

There are just so many ways this could go wrong. 

A simple example is that a WebKitTestRunner that attempts to print out delegate callbacks for unit testing might randomly succeed or fail based on extra delegate callbacks triggered by a totally unrelated test case.

At a minimum, this design is inefficient.
True, we might optimize it if UI process only sends message to web processes that has queried permission or has permission observers.

At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.
Not sure why this would happen. What are they trying to assert or check? I think the messages are handled by WebKit.

> > I think we should reformulate this API to be per-WebView, and have the client app invoke the API on the WebViews it wishes to include.
> The reason we don't do this is client cannot notify SharedWorker and ServiceWorker about the change.
> For permission in these workers, we pick the most "reasonable" WebView to query permission. If we take WebView as parameter, when the permission changes for the WebView, we don't know if we should update state in the workers.

Can WebKit infer which workers to notify based on which WebViews were notified?